src/Core/System/SalesChannel/Validation/SalesChannelValidator.php

```diff
@@ -169,97 +168,68 @@ private function validateLanguages(array $mapping, PreWriteValidationEvent $even
 
         foreach ($mapping as $id => $channel) {
             if (isset($channel['inserts'])) {
-                if (!$this->validInsertCase($channel)) {
+                if (isset($channel['new_default']) && $this->isInvalidInsertCase($channel)) {
                     $inserts[$id] = $channel['new_default'];
                 }
 
                 $duplicatedIds = $this->getDuplicates($channel);
 
-                if ($duplicatedIds) {
+                if ($duplicatedIds !== []) {
                     $duplicates[$id] = $duplicatedIds;
                 }
             }
 
-            if (isset($channel['deletions']) && !$this->validDeleteCase($channel)) {
+            if (isset($channel['deletions'], $channel['current_default']) && $this->isInvalidDeleteCase($channel)) {
                 $deletions[$id] = $channel['current_default'];
             }
 
-            if (isset($channel['updateId']) && !$this->validUpdateCase($channel)) {
+            if (isset($channel['updateId']) && $this->isInvalidUpdateCase($channel)) {
                 $updates[$id] = $channel['updateId'];
             }
         }
 
-        $this->writeInsertViolationExceptions($inserts, $event);
         $this->writeDuplicateViolationExceptions($duplicates, $event);
-        $this->writeDeleteViolationExceptions($deletions, $event);
-        $this->writeUpdateViolationExceptions($updates, $event);
+        $this->writeViolationExceptions($inserts, self::INSERT_VALIDATION_MESSAGE, self::INSERT_VALIDATION_CODE, $event);
+        $this->writeViolationExceptions($deletions, self::DELETE_VALIDATION_MESSAGE, self::DELETE_VALIDATION_CODE, $event);
+        $this->writeViolationExceptions($updates, self::UPDATE_VALIDATION_MESSAGE, self::UPDATE_VALIDATION_CODE, $event);
     }
 
     /**
-     * @param array<string, mixed> $channel
+     * @param array{new_default: string, inserts: list<string>, updateId?: string, deletions?: list<string>, state?: list<string>, current_default?: string} $channel
      */
-    private function validInsertCase(array $channel): bool
+    private function isInvalidInsertCase(array $channel): bool
     {
-        return empty($channel['new_default'])
-            || \in_array($channel['new_default'], $channel['inserts'], true);
+        return !\in_array($channel['new_default'], $channel['inserts'], true);
     }
 
     /**
-     * @param array<string, mixed> $channel
+     * @param array{new_default?: string, inserts?: list<string>, updateId: string, deletions?: list<string>, state?: list<string>, current_default?: string} $channel
      */
-    private function validUpdateCase(array $channel): bool
+    private function isInvalidUpdateCase(array $channel): bool
```

Comment:
Same here, why is this a custom type again instead of referencing one of the above defined ones.

In general this class uses way to many associative arrays, why not simple dump objects 🤯

Reply:
this is different to the above defined types, as we check if some array keys are set before and is therefore different to the defined types. 

I also thought about adding objects. but I already used so much time to understand what is going on here, I didn't want to refactor also that :grimacing: 

But I see, this is still not optimal and not readable, so I will have another look